common/rpc/interceptor/sdk_version.go

```diff
@@ -0,0 +1,97 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package interceptor
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/version/check"
+	"google.golang.org/grpc"
+)
+
+type SDKInfoRecorder interface {
+	RecordSDKInfo(name, version string)
+	GetAndResetSDKInfo() []check.SDKInfo
+}
+
+type SDKVersionInterceptor struct {
+	// Using a sync map to support effienct concurrent updates.
+	// Key type is sdkNameVersion and value type is sdkCount.
+	// Note that we never delete keys from this map as the cardinality of
+	// client versions should be fairly low.
+	sdkInfoCounter sync.Map
+}
+
+type sdkNameVersion struct {
+	name    string
+	version string
+}
+
+type sdkCount struct {
+	count int64
```

Comment:
While it'll likely never be reached anyways, `uint64` gives you twice the counts with the same mem usage since you'll never be negative

Reply:
we should not have this counter. Just use set instead of map for sdk_name_and_version.

Reply:
Can you clarify (maybe w/ an example)? How do you atomically update that? I can understand using mutexes if this code was complex but it's not and fits with `sync.Map`'s use case well. You could have a `*uint64` instead of the struct and use atomics, but I think this is clearer.

Reply:
We could keep the counter and just use it as a boolean for `seen` or `not seen`
```go
atomic.SwapInt64(&value.(*sdkCount).count, 1)
```

Reply:
you don't need the count, the fact that key present in the map means it is seen.